dnsproxy: bind dns proxy to localhost only

mhofstetter opened the pull request:
Currently the dns proxy binds to all interfaces.

With this commit, the dns proxy only gets bound to the localhost interfaces. Therefore, up to 4 DNS servers are created (udpv4, tcpv4, udpv6, tcpv6 - depending on the configuration) which all are using the same DNS handler.

Fixes parts of https://github.com/cilium/cilium/issues/23353

```release-note
DNS Proxy binds to loopback interfaces only
```

mhofstetter committed:
dnsproxy: bind dns proxy to localhost only

Currently the dns proxy is bound to all interfaces.

With this commit, the dns proxy only gets bound to the localhost
interfaces. Therefore, up to 4 DNS servers are created (udpv4, tcpv4,
udpv6, tcpv6 - depending on the configuration) which all are using the
same DNS handler.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
dnsproxy: refactor error-propagation in sessionFactory.SetSocketOptions

This commit introduces proper error propagation when errors occur during
sessionFactory.SetSocketOptions.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
iptables: support switching localOnly (IP) in proxy rule redirect

Currently, switching the localOnly property of a proxy rule doesn't get
reflected in the iptables rules. The checks for adding iptable rules don't
include the IP of a rule. Therefore, when upgrading, the old rule remains
in the table. The same applies for the check whether an outdated rule
should be deleted from the table.

This commit fixes this, and adds support for changing the localOnly
property of a proxy rule.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

mhofstetter committed:
dnsproxy: introduce ipfamily (v4 & v6)

This commit introduces the ipfamily type with its current
implementations v4 & v6. This way duplication can be avoided.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

jrajahalme commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -638,60 +634,62 @@ func StartDNSProxy(
 				time.Sleep(ProxyBindRetryInterval)
 			}
-			log.Fatalf("Failed to start %s DNS Proxy on %s", server.Net, server.Addr)
+			log.Fatalf("Failed to start the %s DNS proxy on %s", server.Net, server.Addr)
```
Wondering if this fatal is executed if ActivateAndServe ever returns without an error? Does it?

mhofstetter replied:
@jrajahalme good catch, thanks! 

indeed this would have failed fatally in the past in case of a proper shutdown. 

even though i assume that in most cases the existing 5s sleep after the return of `ActivateAndServe` would have prevented the occurrence because the application would have shutdown in the meantime anyway

-> fixed

jrajahalme reviewed (approved):
LGTM

mhofstetter committed:
dnsproxy: fixed error handling during successful dns proxy shutdown

Currently, dns server goroutine would exit with log.fatal in case of
returning without an error from ActivateAndServe (in case of a properly
initiated shutdown).

This commit changes this behaviour to only fail fatally in case of a
returned error.

Signed-off-by: Marco Hofstetter <marco.hofstetter@isovalent.com>

pippolo84 commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -1083,42 +1083,93 @@ func ExtractMsgDetails(msg *dns.Msg) (qname string, responseIPs []net.IP, TTL ui
-			if conn != nil {
-				conn.Close()
-			}
+			shutdownServers(dnsServers)
```
This won't have any effect, since `shutdownServers` returns a function.
I suggest to change `shutdownServers` to actually call `Shutdown` on each server and wrap it in a bare `func()` only when assigned to `unbindAddress` to be called later during cleanup.

mhofstetter replied:
@pippolo84 thanks a lot! fixed!

pippolo84 reviewed (changes requested):
Nice one, thanks! :rocket: 
I've just left a comment inline regarding the servers shutdown.

pippolo84 reviewed (approved):
Great, thanks! :100:
